go/trace/trace.go

```diff
@@ -79,35 +88,74 @@ func CopySpan(parentCtx, spanCtx context.Context) context.Context {
+}
+
+func AddGrpcServerOptions(addInterceptors func(s grpc.StreamServerInterceptor, u grpc.UnaryServerInterceptor)) {
+	spanFactory.AddGrpcServerOptions(addInterceptors)
+}
+
+// TracingService is an interface for creating spans or extracting them from Contexts.
+type TracingService interface {
+	// New creates a new span from an existing one, if provided. The parent can also be nil
+	New(parent Span, label string) Span
+
+	// NewClientSpan creates a span for a service call
+	NewClientSpan(parent Span, serviceName, label string) Span
+
+	// Extracts a span from a context, making it possible to annotate the span with additional information.
 	FromContext(ctx context.Context) (Span, bool)
+
+	// Creates a new context containing the provided span
 	NewContext(parent context.Context, span Span) context.Context
-}
 
-var spanFactory SpanFactory = fakeSpanFactory{}
+	// Allows a tracing system to add interceptors to grpc server traffic
+	AddGrpcServerOptions(addInterceptors func(s grpc.StreamServerInterceptor, u grpc.UnaryServerInterceptor))
+
+	// Allows a tracing system to add grpc configuration to server traffic
+	GetGrpcServerOptions() []grpc.ServerOption
 
-// RegisterSpanFactory should be called by a plugin during init() to install a
-// factory that creates Spans for that plugin's tracing framework. Each call to
-// RegisterSpanFactory will overwrite any previous setting. If no factory is
-// registered, the default fake factory will produce Spans whose methods are all
-// no-ops.
-func RegisterSpanFactory(sf SpanFactory) {
-	spanFactory = sf
+	// Allows a tracing system to add grpc configuration to client traffic
+	GetGrpcClientOptions() []grpc.DialOption
 }
 
-type fakeSpanFactory struct{}
+type TracerFactory func(serviceName string) (TracingService, io.Closer, error)
+
+// tracingBackendFactories should be added to by a plugin during init() to install itself
+var tracingBackendFactories = make(map[string]TracerFactory)
 
-func (fakeSpanFactory) New(parent Span) Span                                         { return fakeSpan{} }
-func (fakeSpanFactory) FromContext(ctx context.Context) (Span, bool)                 { return nil, false }
-func (fakeSpanFactory) NewContext(parent context.Context, span Span) context.Context { return parent }
+var spanFactory TracingService = fakeSpanFactory{}
 
-// fakeSpan implements Span with no-op methods.
-type fakeSpan struct{}
+var (
+	tracingServer = flag.String("tracer", "noop", "tracing service to use.")
+)
+
+// StartTracing enables tracing for a named service
+func StartTracing(serviceName string) io.Closer {
+	factory, ok := tracingBackendFactories[*tracingServer]
+	if !ok {
+		options := make([]string, len(tracingBackendFactories))
+		for k := range tracingBackendFactories {
+			options = append(options, k)
+		}
+
+		altStr := strings.Join(options, ", ")
+		log.Error(vterrors.Errorf(vtrpc.Code_INVALID_ARGUMENT, "no such tracing service found. alternatives are: %v", altStr))
```

Comment:
To my taste, the use of `vterrors.Errorf` here feels unnecessary. An alternative might be something like:

`log.Errorf("skipping initialization of unknown tracing service: %v. Valid options: %v", *tracingServer,  strings.Join(options, ", "))`